Show info box if a new version is available on GitHub

bonomat opened the pull request:
Resolves #2233

Note: our latest version 0.4.20.1 is not a valid semver version hence, I added a small logic which uses `coerce` to parse the version string. The resulting version of `0.4.20.1` will be `0.4.20`.  

<img width="1506" alt="image" src="https://user-images.githubusercontent.com/224613/174686044-3361cc0e-454a-4414-9700-5e390f16e838.png">

bonomat committed:
Get daemon version via `api/version`

bonomat committed:
Add infobox to ui

We fetch the daemon version and the latest version from Github and compare them with each other. If the daemon version is outdated a small info box is displayed telling the user that he can upgrade

klochowicz commented on taker-frontend/src/App.tsx:
```diff
@@ -68,10 +79,47 @@ export const BG_LIGHT = "gray.50";
+    if (githubVersion?.tag_name) {
+        let version = semver.parse(githubVersion?.tag_name);
+        if (!version) {
+            // needed because 0.20.1 is not a valid semver version
```
```suggestion
            // needed because 0.4.20.1 is not a valid semver version
```

I believe you would have the same problems with the usual vergen-created version that happens when we're running from master? e.g. `0.4.20-a243bc` (we append git tag to distinguish the versions)
note: the latter is a valid version, but it would fail at comparison, as semver treats such versions as *prereleases* 😕 (edit: the screenshot shows this issue, where it tries to tell the user to downgrade back to `0.4.20`)

bonomat replied:
If you look at the screenshot, it was able to parse the tag correctly as `0.4.20-217-g4d...`

The screenshot was a hack to get it up, i.e. I turned the condition temporarily around 😬 Otherwise it would not show it.

klochowicz replied:
> If you look at the screenshot, it was able to parse the tag correctly as `0.4.20-217-g4d...`
> 
> The screenshot was a hack to get it up, i.e. I turned the condition temporarily around 😬 Otherwise it would not show it.

really? that's good! in the rust `semver` crate the comparison would certainly show it (it treats the tag as prerelease, which should happen before the official release).

klochowicz commented on taker-frontend/src/App.tsx:
```diff
@@ -68,10 +79,47 @@ export const BG_LIGHT = "gray.50";
 export const BG_DARK = "gray.800";
 export const FAQ_URL = "http://faq.itchysats.network";
 
+const fetchGithubVersion = async (setGithubVersion: (version: SemVer | null) => void) => {
```
perhaps we could put `fetchGithubVersion` and `fetchDaemonVersion` into a different file? they clutter the top-level file a bit...

bonomat replied:
can do

klochowicz reviewed (approved):
great addition!

bonomat committed:
Use coerce to parse version from Github

Our latest version on Github is 0.4.20.1 which is not a valid semver version.
Hence, we use `coerce` to parse the version. The resulting version is `0.4.20`, i.e. it drops the last `.1`.
